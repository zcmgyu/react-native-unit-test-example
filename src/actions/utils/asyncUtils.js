/**
 * @param baseName action base name
 * @return {object} an object with the three async action names
 */
export const asyncActionNames = baseName => ({
  failure: `${baseName}_FAILURE`,
  request: `${baseName}_REQUEST`,
  success: `${baseName}_SUCCESS`
});

/**
 * @param actionName {object}
 * @return {object} the three async object actions
 */
export const buildAsyncActions = actionName => ({
  request: () => ({
    type: actionName.request
  }),
  failure: error => ({
    type: actionName.failure,
    error
  }),
  success: data => ({
    type: actionName.success,
    data
  })
});
