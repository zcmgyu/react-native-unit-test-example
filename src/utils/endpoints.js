export const repositories = `https://api.github.com/search/repositories?q=react-native&sort=stars&order=desc`;
