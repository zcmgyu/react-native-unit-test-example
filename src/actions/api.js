import { asyncActionNames, buildAsyncActions } from './utils/asyncUtils.js';

// Constants
export const LOAD_REPOS = asyncActionNames('LOAD_REPOS');

// Actions
export const receiveRepos = buildAsyncActions(LOAD_REPOS);

// Async actions
export const fetchData = (endpoint, fetchAction) => async dispatch => {
  try {
    dispatch(fetchAction.request());
    const response = await fetch(endpoint);
    const responseJson = await response.json();
    dispatch(fetchAction.success(responseJson.items));
  } catch (error) {
    dispatch(fetchAction.failure(error));
  }
};
