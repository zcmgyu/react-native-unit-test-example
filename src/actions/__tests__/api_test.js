import mockStore from 'redux-mock-store';

import { fetchData } from '../api';
import { asyncActionNames, buildAsyncActions } from '../utils/asyncUtils';

const store = mockStore();
const TEST_API = asyncActionNames('TEST_API');
const receiveTestData = buildAsyncActions(TEST_API);

beforeEach(() => {
  store.clearActions();
});

it('should handle TEST_API_SUCCESS action', async () => {
  const response = '{"items": [{"id": 1}]}';
  fetch.mockResponseSuccess(response);
  await store.dispatch(fetchData('/test', receiveTestData));
  expect(store.getActions()).toMatchSnapshot();
});

it('should handle TEST_API_FAILURE action', async () => {
  const mockError = new Error('Look ma! I am an error.');
  fetch.mockResponseFailure(mockError);
  await store.dispatch(fetchData('/test', receiveTestData));
  expect(store.getActions()).toMatchSnapshot();
});
