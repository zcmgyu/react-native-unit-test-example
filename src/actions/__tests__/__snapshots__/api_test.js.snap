// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle TEST_API_FAILURE action 1`] = `
Array [
  Object {
    "type": "TEST_API_REQUEST",
  },
  Object {
    "error": [Error: Look ma! I am an error.],
    "type": "TEST_API_FAILURE",
  },
]
`;

exports[`should handle TEST_API_SUCCESS action 1`] = `
Array [
  Object {
    "type": "TEST_API_REQUEST",
  },
  Object {
    "data": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "TEST_API_SUCCESS",
  },
]
`;
